SCons/Tool/ninja/NinjaState.py

```diff
@@ -0,0 +1,708 @@
+
+    # pylint: disable=too-many-branches
+    def handle_list_action(self, node, action):
+        """TODO write this comment"""
+        results = [
+            self.action_to_ninja_build(node, action=act)
+            for act in action.list
+            if act is not None
+        ]
+        results = [
+            result for result in results if result is not None and result["outputs"]
+        ]
+        if not results:
+            return None
+
+        # No need to process the results if we only got a single result
+        if len(results) == 1:
+            return results[0]
+
+        all_outputs = list({output for build in results for output in build["outputs"]})
+        dependencies = list({dep for build in results for dep in build["implicit"]})
+
+        if results[0]["rule"] == "CMD" or results[0]["rule"] == "GENERATED_CMD":
+            cmdline = ""
+            for cmd in results:
+
+                # Occasionally a command line will expand to a
+                # whitespace only string (i.e. '  '). Which is not a
+                # valid command but does not trigger the empty command
+                # condition if not cmdstr. So here we strip preceding
+                # and proceeding whitespace to make strings like the
+                # above become empty strings and so will be skipped.
+                cmdstr = cmd["variables"]["cmd"].strip()
+                if not cmdstr:
+                    continue
+
+                # Skip duplicate commands
+                if cmdstr in cmdline:
+                    continue
+
+                if cmdline:
+                    cmdline += " && "
+
+                cmdline += cmdstr
+
+            # Remove all preceding and proceeding whitespace
+            cmdline = cmdline.strip()
+
+            # Make sure we didn't generate an empty cmdline
+            if cmdline:
+                ninja_build = {
+                    "outputs": all_outputs,
+                    "rule": get_rule(node, "GENERATED_CMD"),
+                    "variables": {
+                        "cmd": cmdline,
+                        "env": get_command_env(node.env if node.env else self.env),
+                    },
+                    "implicit": dependencies,
+                }
+
+                if node.env and node.env.get("NINJA_POOL", None) is not None:
+                    ninja_build["pool"] = node.env["pool"]
+
+                return ninja_build
+
+        elif results[0]["rule"] == "phony":
+            return {
+                "outputs": all_outputs,
+                "rule": "phony",
+                "implicit": dependencies,
+            }
+
+        elif results[0]["rule"] == "INSTALL":
+            return {
+                "outputs": all_outputs,
+                "rule": get_rule(node, "INSTALL"),
+                "inputs": get_inputs(node),
+                "implicit": dependencies,
+            }
+
+        raise Exception("Unhandled list action with rule: " + results[0]["rule"])
```

Comment:
maybe InternalError? Or ise it a UserError?

Reply:
Fixed in next push.